### PR TITLE
v1

### PR DESCRIPTION
**Pull request by qhwa**
This is the first working version of Hyder Product Server.

status: working-in-progress

### Tasks

* [x] core functions
    * [x] product
    * [x] package
        * [x] build package zip ball
    * [x] rollout
    * [x] policy

* [x] product API
    * [x] list
    * [x] create
    * [x] update
    * [ ] ~~delete~~ (differed)
    * [x] show

* [x] package API
    * [x] create or update
    * [x] list
    * [x] delete
    * [x] show
    * [x] download archive zip
    * [ ] ~~generating update zip balls~~ (differed)

* [x] rollout API
    * [x] create
    * [x] rollback
    * [x] show
    * [x] update a rollout
    * [x] get current rollout of a product
    * [x] list rollouts of a product

* [x] deployment friendly
* [x] keep repo clean by removing unsed files

### CODE REVIEW

File: lib/hps/core.ex [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+          {:error, changeset}
+        end
+
+      other ->
+        other
+    end
+  end
+
+  defp version_conflict?(%{errors: errors}) when is_list(errors) do
+    match?(
```
Comment (bencode) [inline lib/hps/core.ex]: 这个真好用。

File: lib/hps/core.ex [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+
+    package
+    |> Repo.preload([:product, :files])
+    |> Package.update_changeset(attrs)
+    |> Repo.update()
+    |> case do
+      {:ok, package} ->
+        :ok =
+          Storage.save_archive(package)
+          |> refresh_store()
```
Comment (bencode) [inline lib/hps/core.ex]: 我记得这样credo会警告的

File: lib/hps_web/plug_auth.ex [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  @whitelist Application.get_env(:hps, :basic_auth_wl, [])
+             |> Enum.map(fn {u, pwd} -> u <> "\0" <> pwd end)
+
+  case @whitelist do
+    [] ->
+      def authorize_user(conn, _, _), do: conn
+
+    _ ->
+      def authorize_user(conn, user, pwd) do
+        if Enum.any?(@whitelist, &Plug.Crypto.secure_compare(&1, user <> "\0" <> pwd)) do
```
Comment (bencode) [inline lib/hps_web/plug_auth.ex]: 顺便看了篇文章
https://codahale.com/a-lesson-in-timing-attacks/

File: lib/hps/core.ex (hunk 2 above)
Comment (qhwa) [reply lib/hps/core.ex]: 新版本的已经不会了

**Pull request merged**